SYCL/UserDefinedReductions/user_defined_reductions.cpp

```diff
@@ -0,0 +1,205 @@
+
+template <typename T> struct UserDefinedBitAnd {
+  T operator()(const T &a, const T &b) const { return a & b; }
+};
+
+template <typename T> struct UserDefinedMultiplies {
+  T operator()(const T &a, const T &b) const { return a * b; }
+};
+
+struct custom_type_nested {
+  static constexpr int default_i_value = 6;
+  static constexpr float default_f_value = 1.5;
+
+  constexpr custom_type_nested() = default;
+  constexpr custom_type_nested(int i, float f) : i(i), f(f) {}
+
+  int i = default_i_value;
+  float f = default_f_value;
+};
+
+inline bool operator==(const custom_type_nested &lhs,
+                       const custom_type_nested &rhs) {
+  return lhs.i == rhs.i && lhs.f == rhs.f;
+}
+
+inline std::ostream &operator<<(std::ostream &out,
+                                const custom_type_nested &v) {
+  return out << "custom_type_nested { .i = " << v.i << ", .f = " << v.f << "}";
+}
+
+struct custom_type {
+  static constexpr unsigned long long default_ull_value = 42;
+
+  constexpr custom_type() = default;
+  constexpr custom_type(int i, float f, unsigned long long ull)
+      : n(i, f), ull(ull) {}
+
+  custom_type_nested n;
+  unsigned long long ull = default_ull_value;
+};
+
+inline bool operator==(const custom_type &lhs, const custom_type &rhs) {
+  return lhs.n == rhs.n && lhs.ull == rhs.ull;
+}
+
+inline custom_type operator+(const custom_type &lhs, const custom_type &rhs) {
+  return custom_type(lhs.n.i + rhs.n.i, lhs.n.f + rhs.n.f, lhs.ull + rhs.ull);
+}
+
+struct custom_type_wo_default_ctor {
+  static constexpr unsigned long long default_ull_value = 42;
+
+  constexpr custom_type_wo_default_ctor() = delete;
+  constexpr custom_type_wo_default_ctor(int i, float f, unsigned long long ull)
+      : n(i, f), ull(ull) {}
+
+  custom_type_nested n;
+  unsigned long long ull = default_ull_value;
+};
+
+inline bool operator==(const custom_type_wo_default_ctor &lhs,
+                       const custom_type_wo_default_ctor &rhs) {
+  return lhs.n == rhs.n && lhs.ull == rhs.ull;
+}
+
+inline custom_type_wo_default_ctor
+operator+(const custom_type_wo_default_ctor &lhs,
+          const custom_type_wo_default_ctor &rhs) {
+  return custom_type_wo_default_ctor(lhs.n.i + rhs.n.i, lhs.n.f + rhs.n.f,
+                                     lhs.ull + rhs.ull);
+}
+
+template <typename T, std::size_t... Is>
+constexpr std::array<T, sizeof...(Is)> init_array(T value,
+                                                  std::index_sequence<Is...>) {
+  return {{(static_cast<void>(Is), value)...}};
+}
+
+using namespace sycl;
+
+template <typename InputContainer, typename OutputContainer,
+          class BinaryOperation>
+void test(queue q, InputContainer input, OutputContainer output,
+          BinaryOperation binary_op,
+          typename OutputContainer::value_type identity,
+          typename OutputContainer::value_type init) {
+  typedef typename InputContainer::value_type InputT;
+  typedef typename OutputContainer::value_type OutputT;
```

Comment:
Prefer `using AliasTy = <...>;`

Reply:
Applied in https://github.com/intel/llvm-test-suite/pull/1395/commits/597687b80c40d73cd75223c9c800ef96ff841e32